Client/Frontend/Browser/Tabs/RemoteTabs/RemoteTabsPanelMiddleware.swift

```diff
@@ -0,0 +1,71 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import UIKit
+import Storage
+import Common
+import Shared
+import Redux
+
+class RemoteTabsPanelMiddleware {
+    private let profile: Profile
+
+    init(profile: Profile = AppContainer.shared.resolve()) {
+        self.profile = profile
+    }
+
+    lazy var remoteTabsPanelProvider: Middleware<AppState> = { state, action in
+        switch action {
+        case RemoteTabsPanelAction.refreshTabs:
+            self.refreshTabs(updateCache: true)
+        case RemoteTabsPanelAction.refreshCachedTabs:
+            self.refreshTabs(updateCache: false)
+        default:
+            break
+        }
+    }
+
+    // MARK: - Internal Utilities
+
+    private func refreshTabs(updateCache: Bool = false) {
+        ensureMainThread { [self] in
+            guard profile.hasSyncableAccount() else {
+                DispatchQueue.main.async {
+                    store.dispatch(RemoteTabsPanelAction.refreshDidFail(.notLoggedIn))
+                }
+                return
+            }
+
+            let syncEnabled = (profile.prefs.boolForKey(PrefsKeys.TabSyncEnabled) == true)
+            guard syncEnabled else {
+                DispatchQueue.main.async {
+                    store.dispatch(RemoteTabsPanelAction.refreshDidFail(.syncDisabledByUser))
+                }
+                return
+            }
+
+            profile.getCachedClientsAndTabs().uponQueue(.main) { [weak self] result in
```

Comment:
There is a wrapper of this call without Deferred that we should use.
Note: We have an undergoing effort to remove the usage of Deferred in this case we can't remove yet we might be able to do it after the refactoring if is only used in the legacy version 🤞

Reply:
@yoanarios I updated to use the wrapper in https://github.com/mozilla-mobile/firefox-ios/pull/16740/commits/54fc808ffbae04b528a967d0f0474ebda00f0a63. 

I noticed two minor things however: there's only a wrapper for the cached call (I added a 2nd matching wrapper for the call that also syncs tabs), and the wrappers were returning a default empty array (`?? []`) instead of an optional, but in our case we need to know if the result was nil or actually just empty (zero tabs).

LMK if you see anything amiss with those  ☝️ changes however and I can update